NEIGHBORHOODS IN THE CITY OF TORONTO

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


First import some libraries

Download and explore data set

First import urllib and beautiful soup in order to explore data from wikipedia

In [2]:
# import libraries
import urllib.request
#import urllib2
#from bs4 import BeautifulSoup

In [3]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
page = urllib.request.urlopen(url)

In [4]:
from bs4 import BeautifulSoup

In [5]:
soup = BeautifulSoup(page, "lxml")

In [6]:
#print(soup.prettify())
#soup.title

We can check the HTML code from the web page with 'prettify' and
We can verify the title of the web page

In [7]:
all_tables=soup.find_all("table")
#all_tables

 **Now let's find the tables in the wikipedia web page**

In [8]:
right_table=soup.find('table', class_='wikitable sortable')
#right_table

**roll dow until find 'wikitable sortable' it is the table that we need**

In [9]:
A=[]
B=[]
C=[]
#D=[]
#E=[]
 
for row in right_table.findAll('tr'):
    cells=row.findAll('td')
    if len(cells)==3:
        A.append(cells[0].find(text=True))
        B.append(cells[1].find(text=True))
        C.append(cells[2].find(text=True))
        

**Time to extract the content string from within each part of table marked as < td > and add them to the lists from A to C using the -find- option**

In [10]:
df=pd.DataFrame(A,columns=['Postal Code'])
df['Borough']=B
df['Neighbourhood']=C
df.head(15)



Postal Code           Borough     Neighbourhood
0          M1A      Not assigned     Not assigned

1          M2A      Not assigned     Not assigned

2          M3A        North York         Parkwoods
3          M4A        North York  Victoria Village
4          M5A  Downtown Toronto      Harbourfront
5          M5A  Downtown Toronto       Regent Park
6          M6A        North York  Lawrence Heights
7          M6A        North York    Lawrence Manor
8          M7A      Queen's Park     Not assigned

9          M8A      Not assigned     Not assigned

10         M9A         Etobicoke  Islington Avenue
11         M1B       Scarborough             Rouge
12         M1B       Scarborough           Malvern
13         M2B      Not assigned     Not assigned

14         M3B        North York  Don Mills North

In [11]:
df.drop(df.loc[df['Borough']=='Not assigned'].index, inplace=True)
df.head(15)

Postal Code           Borough     Neighbourhood
2          M3A        North York         Parkwoods
3          M4A        North York  Victoria Village
4          M5A  Downtown Toronto      Harbourfront
5          M5A  Downtown Toronto       Regent Park
6          M6A        North York  Lawrence Heights
7          M6A        North York    Lawrence Manor
8          M7A      Queen's Park     Not assigned

10         M9A         Etobicoke  Islington Avenue
11         M1B       Scarborough             Rouge
12         M1B       Scarborough           Malvern
14         M3B        North York  Don Mills North

15         M4B         East York  Woodbine Gardens
16         M4B         East York     Parkview Hill
17         M5B  Downtown Toronto          Ryerson

18         M5B  Downtown Toronto  Garden District

**Now cleaning the data frame removing 'Not assigned' values**

In [12]:
df1 =df.groupby(['Postal Code','Borough'])['Neighbourhood'].apply(lambda x: ', '.join(x.astype(str))).reset_index()
df1.head(15)

Postal Code      Borough                                      Neighbourhood
0          M1B  Scarborough                                     Rouge, Malvern
1          M1C  Scarborough             Highland Creek, Rouge Hill, Port Union
2          M1E  Scarborough                Guildwood\n, Morningside, West Hill
3          M1G  Scarborough                                             Woburn
4          M1H  Scarborough                                        Cedarbrae\n
5          M1J  Scarborough                                Scarborough Village
6          M1K  Scarborough      East Birchmount Park\n, Ionview, Kennedy Park
7          M1L  Scarborough                    Clairlea, Golden Mile, Oakridge
8          M1M  Scarborough  Cliffcrest, Cliffside, Scarborough Village West\n
9          M1N  Scarborough                      Birch Cliff, Cliffside West\n
10         M1P  Scarborough  Dorset Park, Scarborough Town Centre, Wexford ...
11         M1R  Scarborough                                  Maryvale, Wexford
12         M1S  Scarborough                                          Agincourt
13         M1T  Scarborough        Clarks Corners\n, Sullivan\n, Tam O'Shanter
14         M1V  Scarborough  Agincourt North, L'Amoreaux East\n, Milliken, ...

**Using groupby to put the same Neighbourhood associated to a postal code**

In [13]:
df1['Neighbourhood'].replace(['Not assigned\n'], 'Queens Park',inplace=True)
df1.head()

Postal Code      Borough                           Neighbourhood
0         M1B  Scarborough                          Rouge, Malvern
1         M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2         M1E  Scarborough     Guildwood\n, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                             Cedarbrae\n

*OPTIONAL*

In [14]:
# I WANT TO SORT THE DATA FRAME FOR ORGANIZE THE DATA
sort_by_code = df1.sort_values('Postal Code')
sort_by_code.head()

Postal Code      Borough                           Neighbourhood
0         M1B  Scarborough                          Rouge, Malvern
1         M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2         M1E  Scarborough     Guildwood\n, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                             Cedarbrae\n

**The data frame is sorted just for organize the data**

**for the 9th cell in the table on the Wikipedia page, the value of the Borough and the Neighborhood columns will be Queen's Park. For me, after I sorted the data frame the cell is the 85th**

In [15]:
sort_by_code.shape

(103, 3)

**103 ROWS**

**This is the same data frame with another name. **

In [16]:
sortdf1 = df1.sort_values('Postal Code')
sortdf1.head()


Postal Code      Borough                           Neighbourhood
0         M1B  Scarborough                          Rouge, Malvern
1         M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2         M1E  Scarborough     Guildwood\n, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                             Cedarbrae\n

**I decide to use de Geospatial_Coordinates.csv after spend a lot of time trying with geocode. The results were none all the time, or falling in errors
**

**Faster in this way**

In [17]:
dft = pd.read_csv('Geospatial_Coordinates.csv')
dft.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

**Just delete the column Postal code and keep the latitude and longitude columns**

In [18]:
del dft['Postal Code']
dft.head()

Latitude  Longitude
0  43.806686 -79.194353
1  43.784535 -79.160497
2  43.763573 -79.188711
3  43.770992 -79.216917
4  43.773136 -79.239476

**Finally concatenate the two data frames**

In [19]:
dfTN=pd.concat([sortdf1,dft], axis=1)


In [20]:
dfTN.head(15)

Postal Code      Borough  \
0          M1B  Scarborough   
1          M1C  Scarborough   
2          M1E  Scarborough   
3          M1G  Scarborough   
4          M1H  Scarborough   
5          M1J  Scarborough   
6          M1K  Scarborough   
7          M1L  Scarborough   
8          M1M  Scarborough   
9          M1N  Scarborough   
10         M1P  Scarborough   
11         M1R  Scarborough   
12         M1S  Scarborough   
13         M1T  Scarborough   
14         M1V  Scarborough   

                                        Neighbourhood   Latitude  Longitude  
0                                      Rouge, Malvern  43.806686 -79.194353  
1              Highland Creek, Rouge Hill, Port Union  43.784535 -79.160497  
2                 Guildwood\n, Morningside, West Hill  43.763573 -79.188711  
3                                              Woburn  43.770992 -79.216917  
4                                         Cedarbrae\n  43.773136 -79.239476  
5                                 Scarborough Village  43.744734 -79.239476  
6       East Birchmount Park\n, Ionview, Kennedy Park  43.727929 -79.262029  
7                     Clairlea, Golden Mile, Oakridge  43.711112 -79.284577  
8   Cliffcrest, Cliffside, Scarborough Village West\n  43.716316 -79.239476  
9                       Birch Cliff, Cliffside West\n  43.692657 -79.264848  
10  Dorset Park, Scarborough Town Centre, Wexford ...  43.757410 -79.273304  
11                                  Maryvale, Wexford  43.750072 -79.295849  
12                                          Agincourt  43.794200 -79.262029  
13        Clarks Corners\n, Sullivan\n, Tam O'Shanter  43.781638 -79.304302  
14  Agincourt North, L'Amoreaux East\n, Milliken, ...  43.815252 -79.284577

**I put the longitude and latitude of Toronto city**

In [21]:
Latitude = 43.70011 
Longitude = -79.4163

**Using folium and generate the map and the differents points related to lat and long**

revizar

In [22]:
Toronto_map = folium.Map(location=[Latitude, Longitude], zoom_start=15) # generate map centred around Ecco


# add Ecco as a red circle mark
folium.features.CircleMarker(
    [Latitude, Longitude],
    radius=10,
    popup='Ecco',
    fill=True,
    color='red',
    fill_color='red',
    fill_opacity=0.6
    ).add_to(Toronto_map)

#, dfTN.Neighbourhood
# add popular spots to the map as blue circle markers
for Latitude, Longitude in zip(dfTN.Latitude, dfTN.Longitude):
    folium.features.CircleMarker(
        [Latitude, Longitude],
        radius=5,
        popup=label,
        fill=True,
        color='blue',
        fill_color='blue',
        fill_opacity=0.6
        ).add_to(Toronto_map)

# display map
Toronto_map

NameError: name 'label' is not defined

Grouping rows by neighbourhood 

In [23]:
toronto_grouped = dfTN.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood   Latitude  Longitude
0                       Adelaide\n, King\n, Richmond\n  43.650571 -79.384568
1                                            Agincourt  43.794200 -79.262029
2    Agincourt North, L'Amoreaux East\n, Milliken, ...  43.815252 -79.284577
3    Albion Gardens\n, Beaumond Heights, Humbergate...  43.739416 -79.588437
4                               Alderwood, Long Branch  43.602414 -79.543484
5    Bathurst Manor, Downsview North\n, Wilson Heights  43.754328 -79.442259
6                                      Bayview Village  43.786947 -79.385975
7                  Bedford Park, Lawrence Manor East\n  43.733283 -79.419750
8                                          Berczy Park  43.644771 -79.373306
9                        Birch Cliff, Cliffside West\n  43.692657 -79.264848
10   Bloordale Gardens\n, Eringate\n, Markland Wood...  43.643515 -79.577201
11      Brockton\n, Exhibition Place, Parkdale Village  43.636847 -79.428191
12   Business Reply Mail Processing Centre 969 East...  43.662744 -79.321558
13                       CFB Toronto, Downsview East\n  43.737473 -79.464763
14   CN Tower, Bathurst Quay\n, Island airport\n, H...  43.628947 -79.394420
15                         Cabbagetown, St. James Town  43.667967 -79.367675
16                               Caledonia-Fairbanks\n  43.689026 -79.453512
17             Canada Post Gateway Processing Centre\n  43.636966 -79.615819
18                                         Cedarbrae\n  43.773136 -79.239476
19                                Central Bay Street\n  43.657952 -79.387383
20           Chinatown, Grange Park, Kensington Market  43.653206 -79.400049
21                                          Christie\n  43.669542 -79.422564
22                                Church and Wellesley  43.665860 -79.383160
23                     Clairlea, Golden Mile, Oakridge  43.711112 -79.284577
24         Clarks Corners\n, Sullivan\n, Tam O'Shanter  43.781638 -79.304302
25   Cliffcrest, Cliffside, Scarborough Village West\n  43.716316 -79.239476
26   Cloverdale\n, Islington, Martin Grove\n, Princ...  43.650943 -79.554724
27                    Commerce Court, Victoria Hotel\n  43.648198 -79.379817
28                                        Davisville\n  43.704324 -79.388790
29                                  Davisville North\n  43.712751 -79.390197
30   Deer Park, Forest Hill SE\n, Rathnelly, South ...  43.686412 -79.400049
31    Del Ray\n, Keelesdale, Mount Dennis, Silverthorn  43.691116 -79.476013
32            Design Exchange, Toronto Dominion Centre  43.647177 -79.381576
33                                   Don Mills North\n  43.745906 -79.352188
34   Dorset Park, Scarborough Town Centre, Wexford ...  43.757410 -79.273304
35                      Dovercourt Village, Dufferin\n  43.669005 -79.442259
36                                 Downsview Central\n  43.728496 -79.495697
37                               Downsview Northwest\n  43.761631 -79.520999
38                                      Downsview West  43.739015 -79.506944
39              Downsview, North Park\n, Upwood Park\n  43.713756 -79.490074
40       East Birchmount Park\n, Ionview, Kennedy Park  43.727929 -79.262029
41                                        East Toronto  43.685347 -79.338106
42                                    Emery, Humberlea  43.724766 -79.532242
43                    Fairview\n, Henry Farm, Oriole\n  43.778517 -79.346556
44              First Canadian Place, Underground city  43.648429 -79.382280
45                  Flemingdon Park, Don Mills South\n  43.725900 -79.340923
46               Forest Hill North, Forest Hill West\n  43.696948 -79.411307
47                                         Glencairn\n  43.709577 -79.445073
48                 Guildwood\n, Morningside, West Hill  43.763573 -79.188711
49                    Harbord\n, University of Toronto  43.662696 -79.400049
50   Harbourfront East\n, Toronto Islands, Union St...  43.640816 -79.381752
51                           Harbourf

In [24]:
toronto_grouped.shape

(103, 3)

Segment and cluster only the neighborhoods in Toronto. Let's slice the original dataframe and create a new dataframe of the Toronto data.

In [26]:
dfTN.loc[0, 'Neighbourhood']

'Rouge, Malvern'

Get the neighborhood's latitude and longitude values

In [27]:
neighbourhood_latitude = dfTN.loc[0, 'Latitude'] # neighborhood latitude value
neighbourhood_longitude = dfTN.loc[0, 'Longitude'] # neighborhood longitude value

neighbourhood_name = dfTN.loc[0, 'Neighbourhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighbourhood_name, 
                                                               neighbourhood_latitude, 
                                                               neighbourhood_longitude))

Latitude and longitude values of Rouge, Malvern are 43.806686299999996, -79.19435340000001.


get the top 100 venues that are in Rouge, Malvern within a radius of 500 meters

In [29]:

LIMIT = 100 # limit of number of venues returned by Foursquare API

radius = 500 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighbourhood_latitude, 
    neighbourhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=ZFGJMBLROE02FJR2DHPIYB2KE5514IO2VLKCOGSVMPCYFMBD&client_secret=0QIALZFLYEQY4DOQA2I3MYNGH4YMFXOI3ERHDZU5DFT2TSDR&v=20180605&ll=43.806686299999996,-79.19435340000001&radius=500&limit=100'

Examine the results

In [32]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5cd2098a6a60712122f22e3a'},
 'response': {'warning': {'text': "There aren't a lot of results near you. Try something more general, reset your filters, or expand the search area."},
  'headerLocation': 'Malvern',
  'headerFullLocation': 'Malvern, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 1,
  'suggestedBounds': {'ne': {'lat': 43.8111863045, 'lng': -79.18812958073042},
   'sw': {'lat': 43.80218629549999, 'lng': -79.2005772192696}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4bb6b9446edc76b0d771311c',
       'name': "Wendy's",
       'location': {'crossStreet': 'Morningside & Sheppard',
        'lat': 43.80744841934756,
        'lng': -79.19905558052072,
        'labeledLatLngs': [{'label': 'display'

In [33]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now to clean the json and structure it into a pandas dataframe.

In [34]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name            categories        lat        lng
0  Wendy's  Fast Food Restaurant  43.807448 -79.199056

In [35]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

1 venues were returned by Foursquare.


Explore Neighborhoods in Toronto

In [36]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Run the above function on each neighborhood and create a new dataframe called toronto_venues.

In [37]:
toronto_venues = getNearbyVenues(names=dfTN['Neighbourhood'],
                                   latitudes=dfTN['Latitude'],
                                   longitudes=dfTN['Longitude']
                                  )

Rouge, Malvern
Highland Creek, Rouge Hill, Port Union
Guildwood
, Morningside, West Hill
Woburn
Cedarbrae

Scarborough Village
East Birchmount Park
, Ionview, Kennedy Park
Clairlea, Golden Mile, Oakridge
Cliffcrest, Cliffside, Scarborough Village West

Birch Cliff, Cliffside West

Dorset Park, Scarborough Town Centre, Wexford Heights
Maryvale, Wexford
Agincourt
Clarks Corners
, Sullivan
, Tam O'Shanter
Agincourt North, L'Amoreaux East
, Milliken, Steeles East

L'Amoreaux West

Upper Rouge
Hillcrest Village
Fairview
, Henry Farm, Oriole

Bayview Village
Silver Hills
, York Mills
Newtonbrook, Willowdale
Willowdale South

York Mills West

Willowdale West
Parkwoods
Don Mills North

Flemingdon Park, Don Mills South

Bathurst Manor, Downsview North
, Wilson Heights
Northwood Park, York University
CFB Toronto, Downsview East

Downsview West
Downsview Central

Downsview Northwest

Victoria Village
Woodbine Gardens, Parkview Hill
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toront

In [38]:
print(toronto_venues.shape)
toronto_venues.head()

(2243, 7)


Neighbourhood  Neighbourhood Latitude  \
0                          Rouge, Malvern               43.806686   
1  Highland Creek, Rouge Hill, Port Union               43.784535   
2     Guildwood\n, Morningside, West Hill               43.763573   
3     Guildwood\n, Morningside, West Hill               43.763573   
4     Guildwood\n, Morningside, West Hill               43.763573   

   Neighbourhood Longitude                            Venue  Venue Latitude  \
0               -79.194353                          Wendy's       43.807448   
1               -79.160497            Royal Canadian Legion       43.782533   
2               -79.188711  Swiss Chalet Rotisserie & Grill       43.767697   
3               -79.188711                G & G Electronics       43.765309   
4               -79.188711                       Marina Spa       43.766000   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.163085                   Bar  
2       -79.189914           Pizza Place  
3       -79.191537     Electronics Store  
4       -79.191000                   Spa

check how many venues were returned for each neighborhood

In [39]:
toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                                
Adelaide\n, King\n, Richmond\n                                         100   
Agincourt                                                                4   
Agincourt North, L'Amoreaux East\n, Milliken, S...                       3   
Albion Gardens\n, Beaumond Heights, Humbergate\...                       9   
Alderwood, Long Branch                                                  10   
Bathurst Manor, Downsview North\n, Wilson Heights                       19   
Bayview Village                                                          4   
Bedford Park, Lawrence Manor East\n                                     23   
Berczy Park                                                             57   
Birch Cliff, Cliffside West\n                                            4   
Bloordale Gardens\n, Eringate\n, Markland Wood,...                       6   
Brockton\n, Exhibition Place, Parkdale Village                          19   
Business Reply Mail Processing Centre 969 Easte...                      17   
CFB Toronto, Downsview East\n                                            3   
CN Tower, Bathurst Quay\n, Island airport\n, Ha...                      15   
Cabbagetown, St. James Town                                             44   
Caledonia-Fairbanks\n                                                    6   
Canada Post Gateway Processing Centre\n                                 10   
Cedarbrae\n                                                              7   
Central Bay Street\n                                                    88   
Chinatown, Grange Park, Kensington Market                              100   
Christie\n                                                              16   
Church and Wellesley                                                    88   
Clairlea, Golden Mile, Oakridge                                          9   
Clarks Corners\n, Sullivan\n, Tam O'Shanter                             10   
Cliffcrest, Cliffside, Scarborough Village West\n                        2   
Cloverdale\n, Islington, Martin Grove\n, Prince...                       2   
Commerce Court, Victoria Hotel\n                                       100   
Davisville\n                                                            32   
Davisville North\n                                                       7   
Deer Park, Forest Hill SE\n, Rathnelly, South H...                      14   
Del Ray\n, Keelesdale, Mount Dennis, Silverthorn                         4   
Design Exchange, Toronto Dominion Centre                               100   
Don Mills North\n                                                        4   
Dorset Park, Scarborough Town Centre, Wexford H...                       6   
Dovercourt Village, Dufferin\n                                          20   
Downsview Central\n                                                      4   
Downsview Northwest\n                                                    5   
Downsview West                                                           4   
Downsview, North Park\n, Upwood Park\n                                   4   
East Birchmount Park\n, Ionview, Kennedy Park                            4   
East Toronto                                                             4   
Emery, Humberlea                                                         2   
Fairview\n, Henry Farm, Oriole\n                                        66   
First Canadian Place, Underground city                                 100   
Flemingdon Park, Don Mills South\n                                      21   
Forest Hill North, Forest Hill West\n                                    4   
Glencairn\n                                                              4   
Guildwood\n, Morningside, West Hill                                      8   
Harbord\n, University of Toronto                                        34

find out how many unique categories can be curated from all the returned venues

In [40]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 274 uniques categories.


**Analyze Each Neighborhood**

In [41]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Accessories Store  Adult Boutique  \
0                          Rouge, Malvern                  0               0   
1  Highland Creek, Rouge Hill, Port Union                  0               0   
2     Guildwood\n, Morningside, West Hill                  0               0   
3     Guildwood\n, Morningside, West Hill                  0               0   
4     Guildwood\n, Morningside, West Hill                  0               0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Arcade  Argentinian Restaurant  Art Gallery  \
0             0         0       0                       0            0   
1             0         0       0                       0            0   
2             0         0       0                       0            0   
3             0         0       0                       0            0   
4             0         0       0                       0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           0                    0                 0                   0   

   Auto Garage  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  \
0            0              0          0           0           0       0   
1            0              0          0           0           0       0   
2            0              0          0           0           0       0   
3            0              0          0           0           0       0   
4            0              0          0           0           0       0   

   Bank  Bar  Baseball Field  Baseball Stadium  Basketball Stadium  Beach  \
0     0    0               0                 0                   0      0   
1     0    1               0                 0                   0      0   
2     0    0               0                 0                   0      0   
3     0    0               0                 0                   0      0   
4     0    0               0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0                0         0           0                   0          0   
1                0         0           0                   0          0   
2                0         0           0                   0          0   
3                0         0           0                   0          0   
4                0         0           0                   0          0   

   Bistro  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0       0              0          0         0                     0   
1       0              0          0         0                     0   
2       0              0          0         0                     0   
3       0              0          0         0                     0   
4       0              0          0         0        

new data frame size

In [42]:
toronto_onehot.shape

(2243, 275)

Let's group rows by neighbourhood and by taking the mean of the frequency of occurrence of each category

In [43]:
torontoGrouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
torontoGrouped

Neighbourhood  Accessories Store  \
0                      Adelaide\n, King\n, Richmond\n           0.000000   
1                                           Agincourt           0.000000   
2   Agincourt North, L'Amoreaux East\n, Milliken, ...           0.000000   
3   Albion Gardens\n, Beaumond Heights, Humbergate...           0.000000   
4                              Alderwood, Long Branch           0.000000   
5   Bathurst Manor, Downsview North\n, Wilson Heights           0.000000   
6                                     Bayview Village           0.000000   
7                 Bedford Park, Lawrence Manor East\n           0.000000   
8                                         Berczy Park           0.000000   
9                       Birch Cliff, Cliffside West\n           0.000000   
10  Bloordale Gardens\n, Eringate\n, Markland Wood...           0.000000   
11     Brockton\n, Exhibition Place, Parkdale Village           0.000000   
12  Business Reply Mail Processing Centre 969 East...           0.000000   
13                      CFB Toronto, Downsview East\n           0.000000   
14  CN Tower, Bathurst Quay\n, Island airport\n, H...           0.000000   
15                        Cabbagetown, St. James Town           0.000000   
16                              Caledonia-Fairbanks\n           0.000000   
17            Canada Post Gateway Processing Centre\n           0.000000   
18                                        Cedarbrae\n           0.000000   
19                               Central Bay Street\n           0.000000   
20          Chinatown, Grange Park, Kensington Market           0.000000   
21                                         Christie\n           0.000000   
22                               Church and Wellesley           0.000000   
23                    Clairlea, Golden Mile, Oakridge           0.000000   
24        Clarks Corners\n, Sullivan\n, Tam O'Shanter           0.000000   
25  Cliffcrest, Cliffside, Scarborough Village West\n           0.000000   
26  Cloverdale\n, Islington, Martin Grove\n, Princ...           0.000000   
27                   Commerce Court, Victoria Hotel\n           0.000000   
28                                       Davisville\n           0.000000   
29                                 Davisville North\n           0.000000   
30  Deer Park, Forest Hill SE\n, Rathnelly, South ...           0.000000   
31   Del Ray\n, Keelesdale, Mount Dennis, Silverthorn           0.000000   
32           Design Exchange, Toronto Dominion Centre           0.000000   
33                                  Don Mills North\n           0.000000   
34  Dorset Park, Scarborough Town Centre, Wexford ...           0.000000   
35                     Dovercourt Village, Dufferin\n           0.000000   
36                                Downsview Central\n           0.000000   
37                              Downsview Northwest\n           0.000000   
38                                     Downsview West           0.000000   
39             Downsview, North Park\n, Upwood Park\n           0.000000   
40      East Birchmount Park\n, Ionview, Kennedy Park           0.000000   
41                                       East Toronto           0.000000   
42                                   Emery, Humberlea           0.000000   
43                   Fairview\n, Henry Farm, Oriole\n           0.000000   
44             First Canadian Place, Underground city           0.000000   
45                 Flemingdon Park, Don Mills South\n           0.000000   
46              Forest Hill North, Forest Hill West\n           0.000000   
47                                        Glencairn\n           0.000000   
48                Guildwood\n, Morningside, West Hill           0.000000   
49                   Harbord\n, University of Toronto           0.000000   
50  Harbourfront East\n, Toronto Islands, Union St...           0.000000   
51                          Harbourfront, Regent Park           0.000000   
52          

again the new size

In [44]:
torontoGrouped.shape

(100, 275)

Let's print each neighborhood along with the top 5 most common venues

In [46]:
num_top_venues = 5

for hood in torontoGrouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = torontoGrouped[torontoGrouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide
, King
, Richmond
----
                 venue  freq
0          Coffee Shop  0.06
1                 Café  0.05
2      Thai Restaurant  0.04
3  American Restaurant  0.04
4           Steakhouse  0.04


----Agincourt----
                venue  freq
0      Sandwich Place  0.25
1              Lounge  0.25
2      Breakfast Spot  0.25
3        Skating Rink  0.25
4  Miscellaneous Shop  0.00


----Agincourt North, L'Amoreaux East
, Milliken, Steeles East
----
                venue  freq
0          Playground  0.33
1         Coffee Shop  0.33
2                Park  0.33
3  Miscellaneous Shop  0.00
4       Movie Theater  0.00


----Albion Gardens
, Beaumond Heights, Humbergate
, Jamestown, Mount Olive, Silverstone, South Steeles, Thistletown----
                  venue  freq
0         Grocery Store  0.22
1              Pharmacy  0.11
2  Fast Food Restaurant  0.11
3   Fried Chicken Joint  0.11
4           Coffee Shop  0.11


----Alderwood, Long Branch----
            venue  freq
0     

put this into a pandas dataframe

 a function to sort the venues in descending order.

In [47]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

to create the new dataframe and display the top 10 venues for each neighborhood.

In [48]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighbourhood'] = torontoGrouped['Neighbourhood']

for ind in np.arange(torontoGrouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(torontoGrouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                     Adelaide\n, King\n, Richmond\n           Coffee Shop   
1                                          Agincourt        Breakfast Spot   
2  Agincourt North, L'Amoreaux East\n, Milliken, ...                  Park   
3  Albion Gardens\n, Beaumond Heights, Humbergate...         Grocery Store   
4                             Alderwood, Long Branch           Pizza Place   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café            Steakhouse       Thai Restaurant   
1                Lounge        Sandwich Place          Skating Rink   
2            Playground           Coffee Shop           Yoga Studio   
3              Pharmacy           Coffee Shop        Sandwich Place   
4                  Pool                   Pub           Coffee Shop   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0   American Restaurant                 Hotel                   Gym   
1            Donut Shop                 Diner        Discount Store   
2            Donut Shop          Dessert Shop    Dim Sum Restaurant   
3  Fast Food Restaurant   Fried Chicken Joint            Beer Store   
4                   Gym              Pharmacy    Athletics & Sports   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                Bakery                   Bar           Burger Joint  
1               Dog Run      Doner Restaurant            Yoga Studio  
2                 Diner        Discount Store                Dog Run  
3           Pizza Place   Dumpling Restaurant              Drugstore  
4          Skating Rink        Sandwich Place       Department Store

**Cluster Neighborhoods**

Run k-means to cluster the neighborhood into 5 clusters.

In [64]:
# set number of clusters
kclusters = 5

torontoGrouped_clustering = torontoGrouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(torontoGrouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 2, 1, 2, 2, 2, 2, 2, 2, 2], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood

In [66]:
# add clustering labels
neighbourhoods_venues_sorted.insert(0, 'Cluster labels', kmeans.labels_)

toronto_merged = dfTN

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')


toronto_merged.head() 


Postal Code      Borough                           Neighbourhood   Latitude  \
0         M1B  Scarborough                          Rouge, Malvern  43.806686   
1         M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2         M1E  Scarborough     Guildwood\n, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                             Cedarbrae\n  43.773136   

   Longitude  Cluster labels  Cluster Labels  Cluster_Labels  \
0 -79.194353             2.0             0.0             2.0   
1 -79.160497             3.0             0.0             3.0   
2 -79.188711             2.0             0.0             2.0   
3 -79.216917             0.0             0.0             0.0   
4 -79.239476             2.0             0.0             2.0   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0  Fast Food Restaurant             Drugstore    Dim Sum Restaurant   
1                   Bar   Dumpling Restaurant                 Diner   
2        Medical Center        Breakfast Spot   Rental Car Location   
3           Coffee Shop     Korean Restaurant      Insurance Office   
4    Athletics & Sports   Fried Chicken Joint      Hakka Restaurant   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                 Diner        Discount Store               Dog Run   
1        Discount Store               Dog Run      Doner Restaurant   
2    Mexican Restaurant          Intersection           Pizza Place   
3             Drugstore    Dim Sum Restaurant                 Diner   
4       Thai Restaurant                Bakery  Caribbean Restaurant   

  7th Most Common Venue 8th Most Common Venue        9th Most Common Venue  \
0      Doner Restaurant            Donut Shop          Dumpling Restaurant   
1            Donut Shop             Drugstore                  Yoga Studio   
2                   Spa     Electronics Store  Eastern European Restaurant   
3        Discount Store               Dog Run             Doner Restaurant   
4                  Bank      Doner Restaurant                        Diner   

  10th Most Common Venue  
0            College Gym  
1           Dessert Shop  
2    Dumpling Restaurant  
3             Donut Shop  
4         Discount Store

# Visualize the resulting clusters #

The 'rainbow' doesn't works for me. Then I try threshold_scale and it works

In [68]:
# create map
map_clusters = folium.Map(location=[Latitude, Longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.Spectral(np.linspace(0, 1, len(ys)))

#colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

threshold_scale = np.linspace(toronto_merged['Cluster labels'].min(),
                              toronto_merged['Cluster labels'].max(),
                              5, dtype=int)
threshold_scale = threshold_scale.tolist() # change the numpy array to a list
threshold_scale[-1] = threshold_scale[-1] + 1
# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        threshold_scale=threshold_scale,
        fill=True,
        fill_color='YLOrRd',
        fill_opacity=0.7).add_to(map_clusters)      
map_clusters



**Examine the clusters.**

In [72]:
toronto_merged.loc[toronto_merged['Cluster labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster labels  Cluster Labels  Cluster_Labels  \
14       Scarborough             1.0             1.0             1.0   
20        North York             1.0             1.0             1.0   
23        North York             1.0             1.0             1.0   
30        North York             1.0             1.0             1.0   
40         East York             1.0             1.0             1.0   
50  Downtown Toronto             1.0             1.0             1.0   
74              York             1.0             1.0             1.0   
90         Etobicoke             1.0             1.0             1.0   
98              York             1.0             1.0             1.0   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
14                  Park            Playground           Coffee Shop   
20             Cafeteria                  Park            Donut Shop   
23                  Bank                  Park     Electronics Store   
30                  Park               Airport           Snack Place   
40                  Park     Convenience Store         Metro Station   
50                  Park            Playground                 Trail   
74                  Park              Pharmacy         Women's Store   
90                  Park                 River           Yoga Studio   
98                  Park           Yoga Studio             Drugstore   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
14           Yoga Studio            Donut Shop          Dessert Shop   
20          Dessert Shop    Dim Sum Restaurant                 Diner   
23           Yoga Studio    Dim Sum Restaurant                 Diner   
30           Yoga Studio            Donut Shop    Dim Sum Restaurant   
40    Dim Sum Restaurant                 Diner        Discount Store   
50           Yoga Studio      Doner Restaurant      Department Store   
74  Fast Food Restaurant                Market            Donut Shop   
90      Doner Restaurant          Dessert Shop    Dim Sum Restaurant   
98    Dim Sum Restaurant                 Diner        Discount Store   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
14    Dim Sum Restaurant                 Diner        Discount Store   
20        Discount Store               Dog Run      Doner Restaurant   
23        Discount Store               Dog Run      Doner Restaurant   
30                 Diner        Discount Store               Dog Run   
40               Dog Run      Doner Restaurant            Donut Shop   
50          Dessert Shop    Dim Sum Restaurant                 Diner   
74    Dim Sum Restaurant                 Diner        Discount Store   
90                 Diner        Discount Store               Dog Run   
98               Dog Run      Doner Restaurant            Donut Shop   

   10th Most Common Venue  
14                Dog Run  
20              Drugstore  
23             Donut Shop  
30       Doner Restaurant  
40            Yoga Studio  
50         Discount Store  
74                Dog Run  
90             Donut Shop  
98    Dumpling Restaurant

In [73]:
toronto_merged.loc[toronto_merged['Cluster labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster labels  Cluster Labels  Cluster_Labels  \
14       Scarborough             1.0             1.0             1.0   
20        North York             1.0             1.0             1.0   
23        North York             1.0             1.0             1.0   
30        North York             1.0             1.0             1.0   
40         East York             1.0             1.0             1.0   
50  Downtown Toronto             1.0             1.0             1.0   
74              York             1.0             1.0             1.0   
90         Etobicoke             1.0             1.0             1.0   
98              York             1.0             1.0             1.0   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
14                  Park            Playground           Coffee Shop   
20             Cafeteria                  Park            Donut Shop   
23                  Bank                  Park     Electronics Store   
30                  Park               Airport           Snack Place   
40                  Park     Convenience Store         Metro Station   
50                  Park            Playground                 Trail   
74                  Park              Pharmacy         Women's Store   
90                  Park                 River           Yoga Studio   
98                  Park           Yoga Studio             Drugstore   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
14           Yoga Studio            Donut Shop          Dessert Shop   
20          Dessert Shop    Dim Sum Restaurant                 Diner   
23           Yoga Studio    Dim Sum Restaurant                 Diner   
30           Yoga Studio            Donut Shop    Dim Sum Restaurant   
40    Dim Sum Restaurant                 Diner        Discount Store   
50           Yoga Studio      Doner Restaurant      Department Store   
74  Fast Food Restaurant                Market            Donut Shop   
90      Doner Restaurant          Dessert Shop    Dim Sum Restaurant   
98    Dim Sum Restaurant                 Diner        Discount Store   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
14    Dim Sum Restaurant                 Diner        Discount Store   
20        Discount Store               Dog Run      Doner Restaurant   
23        Discount Store               Dog Run      Doner Restaurant   
30                 Diner        Discount Store               Dog Run   
40               Dog Run      Doner Restaurant            Donut Shop   
50          Dessert Shop    Dim Sum Restaurant                 Diner   
74    Dim Sum Restaurant                 Diner        Discount Store   
90                 Diner        Discount Store               Dog Run   
98               Dog Run      Doner Restaurant            Donut Shop   

   10th Most Common Venue  
14                Dog Run  
20              Drugstore  
23             Donut Shop  
30       Doner Restaurant  
40            Yoga Studio  
50         Discount Store  
74                Dog Run  
90             Donut Shop  
98    Dumpling Restaurant

In [74]:
toronto_merged.loc[toronto_merged['Cluster labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster labels  Cluster Labels  Cluster_Labels  \
0         Scarborough             2.0             0.0             2.0   
2         Scarborough             2.0             0.0             2.0   
4         Scarborough             2.0             0.0             2.0   
5         Scarborough             2.0             0.0             2.0   
6         Scarborough             2.0             0.0             2.0   
7         Scarborough             2.0             0.0             2.0   
8         Scarborough             2.0             0.0             2.0   
9         Scarborough             2.0             0.0             2.0   
10        Scarborough             2.0             0.0             2.0   
11        Scarborough             2.0             0.0             2.0   
12        Scarborough             2.0             0.0             2.0   
13        Scarborough             2.0             0.0             2.0   
15        Scarborough             2.0             0.0             2.0   
17         North York             2.0             0.0             2.0   
18         North York             2.0             0.0             2.0   
19         North York             2.0             0.0             2.0   
22         North York             2.0             0.0             2.0   
24         North York             2.0             0.0             2.0   
25         North York             2.0             0.0             2.0   
26         North York             2.0             0.0             2.0   
27         North York             2.0             0.0             2.0   
28         North York             2.0             0.0             2.0   
29         North York             2.0             0.0             2.0   
31         North York             2.0             0.0             2.0   
32         North York             2.0             0.0             2.0   
33         North York             2.0             0.0             2.0   
34         North York             2.0             0.0             2.0   
35          East York             2.0             0.0             2.0   
36          East York             2.0             0.0             2.0   
37       East Toronto             2.0             0.0             2.0   
38          East York             2.0             0.0             2.0   
39          East York             2.0             0.0             2.0   
41       East Toronto             2.0             0.0             2.0   
42       East Toronto             2.0             0.0             2.0   
43       East Toronto             2.0             0.0             2.0   
44    Central Toronto             2.0             0.0             2.0   
45    Central Toronto             2.0             0.0             2.0   
46    Central Toronto             2.0             0.0             2.0   
47    Central Toronto             2.0             0.0             2.0   
48    Central Toronto             2.0             0.0             2.0   
49    Central Toronto             2.0             0.0             2.0   
51   Downtown Toronto             2.0             0.0             2.0   
52   Downtown Toronto             2.0             0.0             2.0   
53   Downtown Toronto             2.0             0.0             2.0   
54   Downtown Toronto             2.0             0.0             2.0   
55   Downtown Toronto             2.0             0.0             2.0   
56   Downtown Toronto             2.0             0.0             2.0   
57   Downtown Toronto             2.0             0.0             2.0   
58   Downtown Toronto             2.0             0.0             2.0   
59   Downtown Toronto             2.0             0.0             2.0   
60   Downtown Toronto             2.0             0.0             2.0   
61   Downtown Toronto             2.0             0.0             2.0   
62         North York             2.0             0.0             2.0   
63    Central Toronto             2.0             0.0             2.0   

In [75]:
toronto_merged.loc[toronto_merged['Cluster labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster labels  Cluster Labels  Cluster_Labels  \
1  Scarborough             3.0             0.0             3.0   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
1                   Bar   Dumpling Restaurant                 Diner   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
1        Discount Store               Dog Run      Doner Restaurant   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
1            Donut Shop             Drugstore           Yoga Studio   

  10th Most Common Venue  
1           Dessert Shop

In [76]:
toronto_merged.loc[toronto_merged['Cluster labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster labels  Cluster Labels  Cluster_Labels  \
91   Etobicoke             4.0             0.0             4.0   
97  North York             4.0             0.0             4.0   

         1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
91              Baseball Field           Yoga Studio   Dumpling Restaurant   
97  Construction & Landscaping        Baseball Field           Yoga Studio   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
91                 Diner        Discount Store               Dog Run   
97   Dumpling Restaurant                 Diner        Discount Store   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
91      Doner Restaurant            Donut Shop             Drugstore   
97               Dog Run      Doner Restaurant            Donut Shop   

         10th Most Common Venue  
91  Eastern European Restaurant  
97                    Drugstore